File: bfs/5.longest-palindromic-substring.go
package leetcode

// @leet start
func longestPalindrome(s string) string {
	res := ""
	for i := 0; i <= len(s); i++ {
		res = maxPalid(res, s, i, i)
		res = maxPalid(res, s, i, i+1)
	}
	return res
}

func maxPalid(cur, s string, l, r int) string {
	res := ""
	for l >= 0 && r < len(s) && s[l] == s[r] {
		res = s[l : r+1]
		l--
		r++
	}
	if len(res) > len(cur) {
		return res
	}
	return cur
}

// @leet end

File: BFS/5.longest-palindromic-substring.py
# @leet start
class Solution:
    def maxPal(self, s: str, l: int, r: int, cur: str) -> str:
        res = ""
        while l >= 0 and r < len(s) and s[l] == s[r]:
            res = s[l : r + 1]
            l -= 1
            r += 1
        return res if len(res) > len(cur) else cur

    def longestPalindrome(self, s: str) -> str:
        res = ""
        for i in range(len(s)):
            res = self.maxPal(s, i, i, res)
            res = self.maxPal(s, i, i + 1, res)
        return res


# @leet end
